iroh/src/endpoint/connection.rs

```diff
@@ -220,30 +211,56 @@ pub enum AuthenticationError {
         #[error(std_err)]
         source: ConnectionError,
     },
+    #[error("endpoint state actor stopped")]
+    EndpointStateActorStopped,
+}
+
+impl From<EndpointStateActorStoppedError> for AuthenticationError {
+    #[track_caller]
+    fn from(_value: EndpointStateActorStoppedError) -> Self {
+        e!(Self::EndpointStateActorStopped)
+    }
+}
+
+impl From<EndpointStateActorStoppedError> for ConnectingError {
+    #[track_caller]
+    fn from(_value: EndpointStateActorStoppedError) -> Self {
+        e!(AuthenticationError::EndpointStateActorStopped).into()
+    }
 }
 
 /// Converts a `quinn::Connection` to a `Connection`.
 ///
-/// ## Errors
+/// Returns a [`AuthenticationError`] if the handshake data has not completed,
+/// or if no alpn was set by the remote node.
 ///
-/// Returns a [`AuthenticationError`] if the handshake data has
-/// not completed, or if no alpn was set by the remote node.
+/// Otherwise returns a future that completes once the connection has been registered with the
+/// magicsock. This future can return an [`EndpointStateActorStoppedError`], which will only be
+/// emitted if the endpoint is closing.
+///
+/// The returned future is `'static`, so it can be stored without being lifetime-bound on `&ep`.
 fn conn_from_quinn_conn(
     conn: quinn::Connection,
     ep: &Endpoint,
-) -> Result<Connection, AuthenticationError> {
+) -> Result<
+    impl Future<Output = Result<Connection, EndpointStateActorStoppedError>> + Send + 'static,
+    AuthenticationError,
+> {
```

Comment:
Just curious: is the reason you're not making this an `async fn` because you need to be able to specify the `'static`?  Having it spelled out does make things easier to debug when the 'static bound is lost at some point so I don't think you need to change this.  Just trying to see if I'm understanding things correctly.